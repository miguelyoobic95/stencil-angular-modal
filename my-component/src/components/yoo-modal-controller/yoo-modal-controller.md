---
name: ModalController
category: Core Components
---

The Modal Controller acts as a placeholder for a modal to be inserted. Different modals may be inserted into the same controller. The controller provides the backdrop for that modal so that the content behind it is greyed out. It has no attributes.

The modal is contructed by calling the `generateModal(props: ModalEntry)` method which receives a Modal Entry as an argument.

```yoo-modal-controller.html
    <p>This is a random text in a random page</p>
    <p> Turmoil has engulfed the Galactic Republic. The taxation of trade routes to outlying star systems is in dispute. Hoping to resolve the matter with a blockade of deadly battleships, the greedy Trade Federation has stopped all shipping to the small planet of Naboo. While the Congress of the Republic endlessly debates this alarming chain of events, the Supreme Chancellor has secretly dispatched two Jedi Knights, the guardians of peace and justice in the galaxy, to settle the conflict....</p>

    <yoo-modal-controller></yoo-modal-controller>
    <yoo-button title="Trigger Modal" class="gradient-info" onclick="generate1()"></yoo-button>
    <yoo-button title="Trigger Fullscreen Modal" class="gradient-info" onclick="generate2()"></yoo-button>
    <yoo-button title="Trigger Drawer Modal" class="gradient-info" onclick="generate3()"></yoo-button>
```

```yoo-modal-controller.js

  var modalCtrl = document.querySelector('yoo-modal-controller');
  var modal = document.createElement('yoo-modal');
  modal.title = "Gutten Morgen";
  modal.hasFooter = true;
  modal.classList.add("info");

  var htmlElement = document.createElement('div');
  htmlElement.innerHTML = `<div class="test">This is the test text</div>`;
  var modalProps1 = {title: "Aligato ?", hasFooter : true, content: htmlElement,primaryButtonText: "Confirm", secondaryButtonText: "Cancel"};
  var modalProps2 = {title: "Aligato ?", hasFooter : true, content: htmlElement,primaryButtonText: "Confirm", secondaryButtonText: "Cancel", cssClass: "fullscreen"};
  var modalProps3 = {title: "Aligato ?", hasFooter : true, content: htmlElement,primaryButtonText: "Confirm", secondaryButtonText: "Cancel", cssClass: "drawer"};
  
  function populateShow() {
    modalCtrl.populateModal(modal);
    modal.title = "updated title";
    modal.content = htmlElement;
    modalCtrl.show();
  }

  function generate1() {
    modalCtrl.generateModal(modalProps1);
    console.log('generated', modalCtrl);
    modalCtrl.show();
  }

  function generate2() {
    modalCtrl.generateModal(modalProps2);
    console.log('generated', modalCtrl);
    modalCtrl.show();
  }

  function generate3() {
    modalCtrl.generateModal(modalProps3);
    console.log('generated', modalCtrl);
    modalCtrl.show();
  }

```

## Methods
|Name|Parameters|Description|
|---|---|---|
|`show`|none| Interrupt the current page and display content above|
|`populateModal`|`modal: HTMLElement`| Set the content to be displayed|
|`generateModal`|`props: ModalEntry`| Set the content to be displayed to a Modal created using the ModalEntry parameter|
|`close`|none| Close the modal and placeholder|
