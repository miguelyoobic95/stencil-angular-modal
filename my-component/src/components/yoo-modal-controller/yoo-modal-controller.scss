@import "../../styles/_variables.scss";


:host {
    
    .placeholder {
        z-index: 1000;
        position: absolute;
        top: 0px;
        left: 0px;
        width: 100%;
        height: 100%;
        -webkit-animation: fadeIn .5s cubic-bezier(0.165, 0.840, 0.440, 1.000) forwards;            
        -moz-animation: fadeIn .5s cubic-bezier(0.165, 0.840, 0.440, 1.000) forwards;
        -o-animation: fadeIn .5s cubic-bezier(0.165, 0.840, 0.440, 1.000) forwards;
        animation: fadeIn .5s cubic-bezier(0.165, 0.840, 0.440, 1.000) forwards;
        // background-color: rgba(58, 67, 97, 0.5);
    }
}

:host(.out) {
    -webkit-animation: fadeOut .5s cubic-bezier(0.165, 0.840, 0.440, 1.000) forwards;            
    -moz-animation: fadeOut .5s cubic-bezier(0.165, 0.840, 0.440, 1.000) forwards;
    -o-animation: fadeOut .5s cubic-bezier(0.165, 0.840, 0.440, 1.000) forwards;
    animation: fadeOut .5s cubic-bezier(0.165, 0.840, 0.440, 1.000) forwards;
}

@keyframes fadeIn {
    0% {
      background:rgba(0,0,0,.0);
    }
    100% {
      background:rgba(58, 67, 97, 0.5);
    }
  }

  @keyframes fadeOut {
    0% {
      background:rgba(58, 67, 97, 0.5);
    }
    100% {
      background:rgba(0,0,0,.0);
    }
}