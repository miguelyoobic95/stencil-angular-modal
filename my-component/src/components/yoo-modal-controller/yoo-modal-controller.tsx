import { Component, Element, State, Method, Listen } from '@stencil/core';
import { ModalEntry, setModalProps } from '../yoo-modal/yoo-modal';

@Component({
    tag: 'yoo-modal-controller',
    styleUrl: 'yoo-modal-controller.scss',
    scoped: true
})
export class YooModalControllerComponent {

    @Element() el: HTMLElement;

    @State() modal: HTMLElement = document.createElement('yoo-modal');
    @State() isDisplaying: boolean = false;
    
    @Listen('onClose')
    childClosed() {
        this.el.classList.add('out');
        // setTimeout(() => {this.isDisplaying = false; }, 300);
        this.close();
    }

    @Method()
    show() {
        this.isDisplaying = true;
        this.el.querySelector('div').insertAdjacentElement('afterend', this.modal);
    }

    @Method()
    close() {
        this.el.querySelector('yoo-modal').remove();
        this.isDisplaying = false;        
    }

    @Method()
    populateModal(modal: HTMLElement) {
        this.modal = modal;
    }

    @Method()
    generateModal(props: ModalEntry) {
        let m = document.createElement('yoo-modal');
        m = setModalProps(m, props);
        this.modal = m;
    }

    render() {
        return (
            <div onClick={() => this.close()}class={this.isDisplaying ? 'placeholder' : ''}></div>
        );

    }
}
