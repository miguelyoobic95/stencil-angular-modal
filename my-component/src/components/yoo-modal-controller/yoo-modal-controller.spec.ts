import { render } from '@stencil/core/testing';
import { YooModalControllerComponent } from './yoo-modal-controller';

describe('YooModalControllerComponent', () => {
    it('should build', () => {
        expect(new YooModalControllerComponent()).toBeTruthy();
    });

    describe('Single modal controller', () => {
        fit ('should create a single modal with header and footer', async () => {
            let element = await render({
                components: [YooModalControllerComponent],
                html: '<div><yoo-modal-controller></yoo-modal-controller></div>'
            });
            const modalCtrl = element.querySelector('yoo-modal-controller');
            modalCtrl.generateModal({
                title: 'Gutten Morgen',
                hasFooter: true
            });
            //modalCtrl.show();
            expect(element).toMatchSnapshot();
            //expect(element.querySelector('yoo-modal')).toBeTruthy();
        });
        
    });

});