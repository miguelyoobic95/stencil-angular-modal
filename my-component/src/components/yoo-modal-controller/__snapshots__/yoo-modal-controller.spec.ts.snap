// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooModalControllerComponent Single modal controller should create a single modal with header and footer 1`] = `
<div>
  <yoo-modal-controller
    class="hydrated"
    data-yoo-modal-controller-host=""
  >
    <!---->
    <div
      class=""
      data-yoo-modal-controller=""
    />
  </yoo-modal-controller>
</div>
`;
